File: algorithms/S/Stone Game II/Stone Game II.cpp
class Solution {
public:
    int dp[103][103][2];
    int rec(int i,int m,int p,vector<int>& piles){
        if(i==piles.size()) return 0;
        if(dp[i][m][p]!=-1) return dp[i][m][p];
        int cnt = 0,ans=INT_MIN,n=piles.size();
        for(int j=i;j<min(n,i+2*m);j++){
            cnt += piles[j];
            ans =max(ans, cnt - rec(j+1,max(j-i+1,m),1-p,piles));
        }
        return dp[i][m][p] = ans;
    }
    int stoneGameII(vector<int>& piles) {
        int sum = 0;
        memset(dp,-1,sizeof(dp));
        for(auto i:piles) sum += i;
        return (sum + rec(0,1,0,piles))/2;
    }
};

File: scripts/algorithms/S/Stone Game II/Stone Game II.py
class Solution:
    def stoneGameII(self, piles: List[int]) -> int:
        n = len(piles)
        dp = {} 
        def recursion(index,M):
            # if we reached to the end we cannot score any value
            if index == n:
                return 0
            # we search if we have solved the same case earlier
            if (index,M) in dp:
                return dp[(index,M)] 
            # total remaining score is the sum of array from index to the end
            total = sum(piles[index:])           
            # if we can take the complete array it is the best choice
            if index + 2*M >= n :return total
            # my_score is the score we are getting as the player who is playing
            my_score = 0
            for x in range(index,index+2*M):
                # opponent score will be calculated by next recursion
                opponent_score = recursion(x+1,max(M,x-index+1))
                # my_score is the remaining value of total - opponent_score
                my_score = max(my_score,total - opponent_score)          
            # this is memoization part
            dp[(index,M)] = my_score
            # return the score
            return my_score
        
        return recursion(0,1)
